### ISSUE TITLE
Train with implicit flow

### ISSUE THREAD

**Issue by DanielEggelmann** [sagr4019/ResearchProject#151]
Train neural network based on the implicit flow samples generated by the codegenerator.

**Comment by DanielEggelmann** [comment]
First training with implicit flow. Samples were generated with codegenerator and the parameter `implicit` enables.
Hyperparameters:
- training samples: 18000
- validation samples: 2000
- all other parameters remained the same as earlier trainings

No other changes were made to the code.

Results:
- Training accuracy: ca. 97%
- Epochs: 43

The network seems to recognize any generated data well, I generated around 20 samples for each class, which the network had never seen and they were all classified correctly. However none of the samples crafted by me manually were classified correctly, the same goes for all invalid programs of #147 .
This makes me wonder, if we have a similar problem as with the direct flow. I think we should have a closer look at the generated programs of the codegenerator.

**Comment by DanielEggelmann** [comment]
The trained weights can be found under `models/weights_indirect_flow_iteration1.hdf5`

**Comment by jan-christiansen** [comment]
For the program
```
H x; 
L y; 
y := x
```
I get the following result.
```
Valid:   100.0 %
Invalid:   0.0 %
```
Am I doing something wrong or do you get the same result?

**Comment by DanielEggelmann** [comment]
Yes I do get the same result. The network does not recognize explicit flow anymore, The output seems to be the same for every program not generated, whether valid or invalid.

**Comment by MadeInGer** [comment]
I have checked some valid programs again and can't find any error in the generation.


> For the program
> 
> ```
> H x; 
> L y; 
> y := x
> ```
> 
> I get the following result.
> 
> ```
> Valid:   100.0 %
> Invalid:   0.0 %
> ```
> 
> Am I doing something wrong or do you get the same result?

@jan-christiansen, according to our codegenerator we can't generate such a program. Your program has depth 0, but we can't generate programs with depth 0. The lowest depth of programs we can generate currently is 1, which is a program consisting of e. g. two assignments.

Programs with depth 1 are:

```
Generated 1. valid program regarding implicit flow
securitychecker outputs valid
L x; 
H v; 
L r; 
L d; 
L o; 
H m; 
m := ((x - v) == (m < d)); 
r := o


Generated 2. valid program regarding implicit flow
securitychecker outputs valid
L z; 
H x; 
L a; 
H e; 
while ((z + x) - (271647 + -562871)) do {
    e := ((-63453 + 244347) < (a == 63951))
}
```

The reason lies in line 222 in the codegenerator.

https://github.com/sagr4019/ResearchProject/blob/7f0d26ab7eedce80f9f3e5f3ce3b505d84520f9e/data-generation-and-validation/security-type-system/codegenerator.py#L221-L223

I temporary changed it to generate an integer between `0` and `MAX_DEPTH_COMMAND` and furthermore changed `MAX_DEPTH_COMMAND` to 0 and the results are now programs with depth 0, e. g.:

```
Generated 1. valid program regarding implicit flow
securitychecker outputs valid
H x; 
H v; 
L t; 
H j; 
j := ((t < x) < (j - v))
```

I'm not sure why we/I set the range starting from 1. I think we have to choose one of the following options:
1. do not use programs with depth 0 to test the network. Instead, use programs with at least depth 1
2. set the range in line 222 starting from 0, generate new programs and retrain the network

**Comment by jan-christiansen** [comment]
While it might be good if the generator is able to generate programs of depth zero, my comment was not concerned about this specific form of programs. The trained net does not work for programs of size one as well. I will create a new ticket, in order to investigate the difference between the generated programs and manual written programs.
